### PR TITLE
Fix for TDR caused by fullscreen shadow shader

### PR DESCRIPTION
**Pull request by antonmic**
Issue was somehow ViewSrg::m_directionalLightShadows[m_lightIndex].m_cascadeCount could return a bogus number, causing cascadeCount to blow up the loop that writes to a fixed sized array. Clamped cascadeCount to maxCascadeCount and stored it as a member variable for reference in other functions.

Signed-off-by: antonmic <56370189+antonmic@users.noreply.github.com>

### CODE REVIEW

**Comment by antonmic** [comment]
Screenshot 2022-06-21 155722

**Comment by antonmic** [comment]
^ Parallax is failing in the same way it always does
The multi-render-pipeline failure is something that is unrelated to this change, though it is related to shadows (likely recent CSM pass changes that went in). Will investigate after this.

**Review by santorac — APPROVED** [review]

**Review by dmcdiarmid-ly — APPROVED** [review]
I think it would be a good idea to track down why the cascade count was being set out of bounds.  There's an assert in DirectionalLightFeatureProcessor::SetCascadeCount that should catch it there, so there must be another error somewhere relating to how that count is managed.

File: Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/DirectionalLightShadowCalculator.azsli [hunk 1]
```diff
@@ -149,19 +155,21 @@ void DirectionalShadowCalculator::ComputeShadowCoords()
             m_cascadeIndex = index;
         }
     }
-            
+
+// Todo: No ddx in compute, find a way to leverage world space normal for this
```
Comment (jeremyong-az) [inline Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/DirectionalLightShadowCalculator.azsli:159]: FYI we will have this when we upgrade DXC and enable SM6.6: https://microsoft.github.io/DirectX-Specs/d3d/HLSL_SM_6_6_Derivatives.html#derivative-functions

**Review by jeremyong-az — APPROVED** [review]
Looks good thanks. My best guess as to what might have happened is a period between frames where a draw is occurring but cascades have been disabled and shadow maps unbound which would have caused a fault when we looped up to MaxCascadeCount

**Pull request merged**